--- typo3-src-4.5.0+dfsg1~beta3.orig/t3lib/utility/class.t3lib_utility_command.php
+++ typo3-src-4.5.0+dfsg1~beta3/t3lib/utility/class.t3lib_utility_command.php
@@ -78,12 +78,12 @@
 			// Compile the path & command
 		if ($im_version === 'gm') {
 			$switchCompositeParameters = TRUE;
-			$path = escapeshellarg($path . 'gm' . $isExt) . ' ' . $command;
+			$path = escapeshellcmd($path . 'gm' . $isExt) . ' ' . $command;
 		} else {
 			if ($im_version === 'im6') {
 				$switchCompositeParameters = TRUE;
 			}
-			$path = escapeshellarg($path . (($command == 'composite') ? $combineScript : $command) . $isExt);
+			$path = escapeshellcmd($path . (($command == 'composite') ? $combineScript : $command) . $isExt);
 		}
 
 			// strip profile information for thumbnails and reduce their size
@@ -117,4 +117,4 @@
 
 }
 
-?>
\ No newline at end of file
+?>
