From: Christian Welzel <gawain@camlann.de>
Date: Tue, 13 Nov 2012 12:26:57 +0100
Subject: Fix regression introduced by TYPO3-SA-2012-5

diff --git a/t3lib/class.t3lib_befunc.php b/t3lib/class.t3lib_befunc.php
index 2dd5ced..e5ad43d 100644
--- a/t3lib/class.t3lib_befunc.php
+++ b/t3lib/class.t3lib_befunc.php
@@ -3088,7 +3088,7 @@ final class t3lib_BEfunc {
 			$mainParams .= (t3lib_div::_GET('M') ? '&M=' . rawurlencode(t3lib_div::_GET('M')) : '');
 		}
 
-		$onClick = 'jumpToUrl(' . t3lib_div::quoteJSvalue($script . '?' . $mainParams . $addparams . '&' . $elementName) . '=+(this.checked?1:0),this);';
+		$onClick = 'jumpToUrl(' . t3lib_div::quoteJSvalue($script . '?' . $mainParams . $addparams . '&' . $elementName . '=') . '+(this.checked?1:0),this);';
 
 		return
 		'<input' .
